models/iaf_cond_beta.cpp

```diff
@@ -0,0 +1,521 @@
+  }
+
+  if ( B_.e_ == 0 )
+  {
+    B_.e_ = gsl_odeiv_evolve_alloc( State_::STATE_VEC_SIZE );
+  }
+  else
+  {
+    gsl_odeiv_evolve_reset( B_.e_ );
+  }
+
+  B_.sys_.function = iaf_cond_beta_dynamics;
+  B_.sys_.jacobian = NULL;
+  B_.sys_.dimension = State_::STATE_VEC_SIZE;
+  B_.sys_.params = reinterpret_cast< void* >( this );
+
+  B_.I_stim_ = 0.0;
+}
+
+double
+nest::iaf_cond_beta::get_normalisation_factor( double tau_rise,
+  double tau_decay )
+{
+  // Factor used to normalise the synaptic conductance such that
+  //
+
+  return 1.0 / ( tau_decay * tau_rise );
+}
+
+void
+nest::iaf_cond_beta::calibrate()
+{
+  // ensures initialization in case mm connected after Simulate
+  B_.logger_.init();
+
+  V_.PSConInit_E = nest::iaf_cond_beta::get_normalisation_factor(
+    P_.tau_Erise, P_.tau_Edecay );
+  V_.PSConInit_I = nest::iaf_cond_beta::get_normalisation_factor(
+    P_.tau_Irise, P_.tau_Idecay );
+  V_.RefractoryCounts = Time( Time::ms( P_.t_ref ) ).get_steps();
+
+  // since t_ref >= 0, this can only fail in error
+  assert( V_.RefractoryCounts >= 0 );
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::iaf_cond_beta::update( Time const& origin,
+  const long from,
+  const long to )
+{
+
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+
+    double t = 0.0;
+
+    // numerical integration with adaptive step size control:
+    // ------------------------------------------------------
+    // gsl_odeiv_evolve_apply performs only a single numerical
+    // integration step, starting from t and bounded by step;
+    // the while-loop ensures integration over the whole simulation
+    // step (0, step] if more than one integration step is needed due
+    // to a small integration step size;
+    // note that (t+IntegrationStep > step) leads to integration over
+    // (t, step] and afterwards setting t to step, but it does not
+    // enforce setting IntegrationStep to step-t; this is of advantage
+    // for a consistent and efficient integration across subsequent
+    // simulation intervals
+    while ( t < B_.step_ )
+    {
+      const int status = gsl_odeiv_evolve_apply( B_.e_,
+        B_.c_,
+        B_.s_,
+        &B_.sys_,             // system of ODE
+        &t,                   // from t
+        B_.step_,             // to t <= step
+        &B_.IntegrationStep_, // integration step size
+        S_.y );               // neuronal state
+      if ( status != GSL_SUCCESS )
+      {
+        throw GSLSolverFailure( get_name(), status );
+      }
+    }
+
+    // refractoriness and spike generation
+    if ( S_.r )
+    { // neuron is absolute refractory
+      --S_.r;
+      S_.y[ State_::V_M ] = P_.V_reset; // clamp potential
```

Comment:
this line must be inside the while: the voltage must stay clamped during all the evolution of the ODE, not just "appear" clamped to the user recording it; only the decrement of ``S_.r`` should be outside

Reply:
This section is identical to the existing model [iaf_cond_alpha.cpp](https://github.com/nest/nest-simulator/blob/master/models/iaf_cond_alpha.cpp) so i am not too familiar with it. if there are any issues here they should also be corrected there:

Reply:
@Silmathoron I think this is fine either way, as `S_.y[ State_::V_M ]` has no influence on the other ODEs in the system. Therefore the results will be the same, as long as it is reset properly at the end of the clamping

Reply:
@Silmathoron @dnao Thanks for finding this! For the `iaf_cond_{alpha,beta}` models it makes indeed no difference as @janhahne points out, since `V_m` only enters into the ODE for `V_m`. But to do it really systematically, one needs to clamp inside the `..._dynamics` function. See `ht_neuron` for an example, where correct clamping is indeed required due to non-linear currents. 

@clinssen Would you take note of this in connection with NESTML?

Reply:
@janhahne @heplesser unfortunately that is not true: we are dealing with a conductance-based model here, so in case the ``while ( t < B_.step_ )`` is done in several steps (rather often in my experience), then the synaptic currents are wrong. It will arguably be a small error, but still...

Reply:
@Silmathoron I am sorry, I still dont see it: Could you point me to any line, where a (wrong) value of `S_.y[ State_::V_M ]` enters any variable or property that is used outside of the ODE that describes `S_.y[ State_::V_M ]` (specified by `f[0]`)? Of course the intermediate values of `S_.y[ State_::V_M ]` that are computed during the `while`-loop will be slightly wrong, but as long as they are not used for anything and do not influence any other variables the results at the NEST grid points will be the same as with clamping inside the while-loop. 

However, I completely agree that it would be better to do this systematically inside the `..._dynamics` function, as it becomes important once `S_.y[ State_::V_M ]` is also a part of other ODEs in the ODE-system.

Reply:
No, I did not mean outside of the ODE, but the computation of [``I_syn_ex``](https://github.com/nest/nest-simulator/pull/1038/files/9e982021d3d9684f77c75944e304c9ca5f9dd149#diff-76587cc1da7f6a4cffaf9ac00c1e944aR100) will be wrong, which will add up to the ``V`` error in the derivative.
However, as you said, these errors should remain small, but the integration is, strictly speaking, incorrect.
Regardless, I think this should be handled in another PR so we can ignore it for now, I'll open a new issue.